File: src/swirui/rendering/affine.py
"""Backend-neutral affine transform math for retained scene geometry."""

from __future__ import annotations

import math
from dataclasses import dataclass

from .geometry import Point, Rect

_DETERMINANT_EPSILON = 1.0e-12
_AXIS_ALIGNMENT_EPSILON = 1.0e-12
_ORIGIN = Point()


@dataclass(frozen=True, slots=True)
class Affine2D:
    """Invertible 2D affine transform expressed in logical-DIP coordinates.

    The matrix maps authored scene coordinates into visual coordinates::

        x' = m11 * x + m12 * y + tx
        y' = m21 * x + m22 * y + ty

    SwirUI keeps this transform independent from layout geometry. That makes it
    suitable for retained visual animation while allowing hit testing to map the
    pointer back into authored coordinates exactly.
    """

    m11: float = 1.0
    m12: float = 0.0
    m21: float = 0.0
    m22: float = 1.0
    tx: float = 0.0
    ty: float = 0.0

    def __post_init__(self) -> None:
        values = (self.m11, self.m12, self.m21, self.m22, self.tx, self.ty)
        if any(not math.isfinite(value) for value in values):
            raise ValueError("Affine2D values must be finite.")
        if abs(self.determinant) <= _DETERMINANT_EPSILON:
            raise ValueError("Affine2D must be invertible.")

    @property
    def determinant(self) -> float:
        return self.m11 * self.m22 - self.m12 * self.m21

    @property
    def is_identity(self) -> bool:
        return (
            self.m11 == 1.0
            and self.m12 == 0.0
            and self.m21 == 0.0
            and self.m22 == 1.0
            and self.tx == 0.0
            and self.ty == 0.0
        )

    @property
    def is_axis_aligned(self) -> bool:
        """Return whether axis-aligned rectangles remain axis-aligned after mapping.

        Translation and independent X/Y scale are axis-aligned. Rotation and shear
        are not. A tiny tolerance absorbs floating-point residue from composed
        transforms such as a mathematically complete turn.
        """

        return (
            abs(self.m12) <= _AXIS_ALIGNMENT_EPSILON
            and abs(self.m21) <= _AXIS_ALIGNMENT_EPSILON
        )

    @classmethod
    def rotation(cls, radians: float, *, origin: Point = _ORIGIN) -> Affine2D:
        """Create a counter-clockwise rotation around ``origin``."""

        angle = float(radians)
        if not math.isfinite(angle):
            raise ValueError("Rotation angle must be finite.")
        if not math.isfinite(origin.x) or not math.isfinite(origin.y):
            raise ValueError("Rotation origin must use finite coordinates.")
        cosine = math.cos(angle)
        sine = math.sin(angle)
        return cls(
            m11=cosine,
            m12=-sine,
            m21=sine,
            m22=cosine,
            tx=origin.x - cosine * origin.x + sine * origin.y,
            ty=origin.y - sine * origin.x - cosine * origin.y,
        )

    @classmethod
    def translation(cls, dx: float, dy: float) -> Affine2D:
        """Create a logical-DIP translation transform."""

        return cls(tx=float(dx), ty=float(dy))

    @classmethod
    def uniform_scale(cls, scale: float, *, origin: Point = _ORIGIN) -> Affine2D:
        """Create a positive uniform scale around ``origin``."""

        normalized = float(scale)
        if not math.isfinite(normalized) or normalized <= 0.0:
            raise ValueError("Affine2D uniform scale must be finite and greater than zero.")
        if not math.isfinite(origin.x) or not math.isfinite(origin.y):
            raise ValueError("Scale origin must use finite coordinates.")
        return cls(
            m11=normalized,
            m22=normalized,
            tx=origin.x * (1.0 - normalized),
            ty=origin.y * (1.0 - normalized),
        )

    def transform_point(self, point: Point) -> Point:
        """Map one authored point into visual coordinates."""

        return Point(
            self.m11 * point.x + self.m12 * point.y + self.tx,
            self.m21 * point.x + self.m22 * point.y + self.ty,
        )

    def inverse_transform_point(self, point: Point) -> Point:
        """Map one visual point back to authored coordinates without allocating a matrix.

        Pointer routing invokes inverse mapping frequently. Computing the two
        coordinates directly preserves the exact ``inverse().transform_point()``
        result while avoiding a temporary ``Affine2D`` object for every clip and
        candidate in the retained hit-test path.
        """

        determinant = self.determinant
        x = point.x - self.tx
        y = point.y - self.ty
        return Point(
            (self.m22 * x - self.m12 * y) / determinant,
            (-self.m21 * x + self.m11 * y) / determinant,
        )

    def inverse(self) -> Affine2D:
        """Return the exact inverse transform."""

        determinant = self.determinant
        inverse_m11 = self.m22 / determinant
        inverse_m12 = -self.m12 / determinant
        inverse_m21 = -self.m21 / determinant
        inverse_m22 = self.m11 / determinant
        return Affine2D(
            m11=inverse_m11,
            m12=inverse_m12,
            m21=inverse_m21,
            m22=inverse_m22,
            tx=-(inverse_m11 * self.tx + inverse_m12 * self.ty),
            ty=-(inverse_m21 * self.tx + inverse_m22 * self.ty),
        )

    def then(self, following: Affine2D) -> Affine2D:
        """Compose this transform with ``following`` in visual order.

        ``a.then(b)`` applies ``a`` first and then ``b``. The explicit ordering
        keeps nested retained transforms deterministic and avoids matrix-order
        ambiguity at call sites.
        """

        return Affine2D(
            m11=following.m11 * self.m11 + following.m12 * self.m21,
            m12=following.m11 * self.m12 + following.m12 * self.m22,
            m21=following.m21 * self.m11 + following.m22 * self.m21,
            m22=following.m21 * self.m12 + following.m22 * self.m22,
            tx=following.m11 * self.tx + following.m12 * self.ty + following.tx,
            ty=following.m21 * self.tx + following.m22 * self.ty + following.ty,
        )

    def transform_rect_bounds(self, rect: Rect) -> Rect:
        """Return the axis-aligned visual bounds enclosing transformed ``rect``."""

        corners = (
            self.transform_point(Point(rect.x, rect.y)),
            self.transform_point(Point(rect.right, rect.y)),
            self.transform_point(Point(rect.right, rect.bottom)),
            self.transform_point(Point(rect.x, rect.bottom)),
        )
        left = min(point.x for point in corners)
        top = min(point.y for point in corners)
        right = max(point.x for point in corners)
        bottom = max(point.y for point in corners)
        return Rect(left, top, right - left, bottom - top)

File: native/src/affine.rs
//! Native affine transform math shared by compositor preparation and GPU backends.
//!
//! The Python SceneGraph owns the public transform API. This module mirrors the
//! same six-float matrix convention inside the Rust core so future GPU submission
//! can consume retained transforms without inventing a second ordering contract.

const DETERMINANT_EPSILON: f32 = 1.0e-6;
const AXIS_ALIGNMENT_EPSILON: f32 = 1.0e-6;

#[derive(Clone, Copy, Debug, PartialEq)]
pub(crate) struct Affine2D {
    pub(crate) m11: f32,
    pub(crate) m12: f32,
    pub(crate) m21: f32,
    pub(crate) m22: f32,
    pub(crate) tx: f32,
    pub(crate) ty: f32,
}

impl Affine2D {
    pub(crate) const IDENTITY: Self = Self {
        m11: 1.0,
        m12: 0.0,
        m21: 0.0,
        m22: 1.0,
        tx: 0.0,
        ty: 0.0,
    };

    pub(crate) fn try_from_slice(values: &[f32]) -> Result<Self, &'static str> {
        if values.len() != 6 {
            return Err("Affine2D requires exactly six floats.");
        }
        if values.iter().any(|value| !value.is_finite()) {
            return Err("Affine2D values must be finite.");
        }
        let transform = Self {
            m11: values[0],
            m12: values[1],
            m21: values[2],
            m22: values[3],
            tx: values[4],
            ty: values[5],
        };
        if transform.determinant().abs() <= DETERMINANT_EPSILON {
            return Err("Affine2D must be invertible.");
        }
        Ok(transform)
    }

    pub(crate) fn determinant(self) -> f32 {
        self.m11 * self.m22 - self.m12 * self.m21
    }

    pub(crate) fn is_identity(self) -> bool {
        self == Self::IDENTITY
    }

    pub(crate) fn is_axis_aligned(self) -> bool {
        self.m12.abs() <= AXIS_ALIGNMENT_EPSILON && self.m21.abs() <= AXIS_ALIGNMENT_EPSILON
    }

    pub(crate) fn then(self, following: Self) -> Self {
        Self {
            m11: following.m11 * self.m11 + following.m12 * self.m21,
            m12: following.m11 * self.m12 + following.m12 * self.m22,
            m21: following.m21 * self.m11 + following.m22 * self.m21,
            m22: following.m21 * self.m12 + following.m22 * self.m22,
            tx: following.m11 * self.tx + following.m12 * self.ty + following.tx,
            ty: following.m21 * self.tx + following.m22 * self.ty + following.ty,
        }
    }

    pub(crate) fn transform_point(self, point: [f32; 2]) -> [f32; 2] {
        [
            self.m11 * point[0] + self.m12 * point[1] + self.tx,
            self.m21 * point[0] + self.m22 * point[1] + self.ty,
        ]
    }

    pub(crate) fn inverse(self) -> Self {
        let determinant = self.determinant();
        let m11 = self.m22 / determinant;
        let m12 = -self.m12 / determinant;
        let m21 = -self.m21 / determinant;
        let m22 = self.m11 / determinant;
        Self {
            m11,
            m12,
            m21,
            m22,
            tx: -(m11 * self.tx + m12 * self.ty),
            ty: -(m21 * self.tx + m22 * self.ty),
        }
    }

    pub(crate) fn transform_rect_bounds(self, rect: [f32; 4]) -> [f32; 4] {
        let x = rect[0];
        let y = rect[1];
        let right = x + rect[2];
        let bottom = y + rect[3];
        let corners = [
            self.transform_point([x, y]),
            self.transform_point([right, y]),
            self.transform_point([right, bottom]),
            self.transform_point([x, bottom]),
        ];
        let mut left = corners[0][0];
        let mut top = corners[0][1];
        let mut visual_right = corners[0][0];
        let mut visual_bottom = corners[0][1];
        for point in corners.into_iter().skip(1) {
            left = left.min(point[0]);
            top = top.min(point[1]);
            visual_right = visual_right.max(point[0]);
            visual_bottom = visual_bottom.max(point[1]);
        }
        [left, top, visual_right - left, visual_bottom - top]
    }

    pub(crate) fn to_gpu_array(self) -> [f32; 6] {
        [self.m11, self.m12, self.m21, self.m22, self.tx, self.ty]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn assert_close(actual: f32, expected: f32) {
        assert!(
            (actual - expected).abs() <= 1.0e-4,
            "expected {expected}, got {actual}"
        );
    }

    #[test]
    fn rejects_invalid_native_contract_payloads() {
        assert_eq!(
            Affine2D::try_from_slice(&[1.0, 0.0, 0.0, 1.0, 0.0]),
            Err("Affine2D requires exactly six floats.")
        );
        assert_eq!(
            Affine2D::try_from_slice(&[1.0, 0.0, 0.0, f32::NAN, 0.0, 0.0]),
            Err("Affine2D values must be finite.")
        );
        assert_eq!(
            Affine2D::try_from_slice(&[1.0, 2.0, 2.0, 4.0, 0.0, 0.0]),
            Err("Affine2D must be invertible.")
        );
    }

    #[test]
    fn composition_matches_python_visual_order() {
        let translation = Affine2D::try_from_slice(&[1.0, 0.0, 0.0, 1.0, 12.0, -5.0])
            .expect("translation should be valid");
        let quarter_turn = Affine2D::try_from_slice(&[0.0, -1.0, 1.0, 0.0, 0.0, 0.0])
            .expect("rotation should be valid");
        let composed = translation.then(quarter_turn);
        let point = composed.transform_point([4.0, 3.0]);
        assert_close(point[0], 2.0);
        assert_close(point[1], 16.0);
    }

    #[test]
    fn inverse_round_trips_visual_points() {
        let transform = Affine2D::try_from_slice(&[0.8, -0.6, 0.6, 0.8, 19.0, -11.0])
            .expect("rotation/translation should be invertible");
        let authored = [132.5, 74.25];
        let visual = transform.transform_point(authored);
        let recovered = transform.inverse().transform_point(visual);
        assert_close(recovered[0], authored[0]);
        assert_close(recovered[1], authored[1]);
    }

    #[test]
    fn transformed_bounds_enclose_rotated_rectangle() {
        let quarter_turn = Affine2D::try_from_slice(&[0.0, -1.0, 1.0, 0.0, 0.0, 0.0])
            .expect("rotation should be valid");
        let bounds = quarter_turn.transform_rect_bounds([10.0, 20.0, 30.0, 40.0]);
        assert_close(bounds[0], -60.0);
        assert_close(bounds[1], 10.0);
        assert_close(bounds[2], 40.0);
        assert_close(bounds[3], 30.0);
    }

    #[test]
    fn classifies_axis_alignment_without_losing_gpu_packing_order() {
        let scale_translate = Affine2D::try_from_slice(&[1.5, 0.0, 0.0, 0.75, 8.0, 9.0])
            .expect("scale/translation should be valid");
        let rotation = Affine2D::try_from_slice(&[0.0, -1.0, 1.0, 0.0, 0.0, 0.0])
            .expect("rotation should be valid");
        assert!(Affine2D::IDENTITY.is_identity());
        assert!(scale_translate.is_axis_aligned());
        assert!(!rotation.is_axis_aligned());
        assert_eq!(
            scale_translate.to_gpu_array(),
            [1.5, 0.0, 0.0, 0.75, 8.0, 9.0]
        );
    }
}
